### ISSUE TITLE
Subdocument validation error - Cast to Array failed for value - .save()

### ISSUE THREAD

**Issue by GregoryNEUT** [Automattic/mongoose#8472]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

Cannot `.save()` changes made on a document due to validation error.

**If the current behavior is a bug, please provide the steps to reproduce.**

I have the following schema : 

```
 {
        // Name of the language
        name: new Schema({
          description: {
            type: String,
          },

          translations: [
            new Schema({
              idLanguage: {
                type: Schema.Types.ObjectId,
                ref: 'Language',
              },

              value: {
                type: String,
              },
            }),
          ],
        }),

        // Acronym that define the language (Ex: FR, EN, ...)
        acronym: {
          type: String,
          required: true,
        },
},
```

I'm getting a document from the database using a `.find()` request, then push new data into the `name.translations` array and `.save()`.

I got the following runtime error : 

> {
>   stringError: { ValidationError: Language validation failed: name.translations: Cast to Array failed for value "[ { idLanguage: 'aaaaa0000000000000000000', value: 'Français' } ]" at path "translations", name: Validation failed: translations: Cast to Array failed for value "[ { idLanguage: 'aaaaa0000000000000000000', value: 'Français' } ]" at path "translations"
>     at ValidationError.inspect (/home/X/back/node_modules/@cotep/database/node_modules/mongoose/lib/error/validation.js:61:24)
>     at formatValue (util.js:399:38)
>     at formatProperty (util.js:800:11)
>     at formatObject (util.js:616:17)
>     at formatValue (util.js:578:18)
>     at inspect (util.js:293:10)
>     at format (util.js:222:18)
>     at Console.log (console.js:130:21)
>     at PromiseCommandPattern.error (/home/X/back/compiled/src/DBClasses/AHaveWordCollection.js:503:19)
>     at PromiseCommandPattern.executeFunctionToExecute$ (/home/X/back/compiled/src/FunctionsDefinitions/PromiseCommandPattern.js:380:57)
>     at tryCatch (/home/X/back/node_modules/regenerator-runtime/runtime.js:45:40)
>     at Generator.invoke [as _invoke] (/home/X/back/node_modules/regenerator-runtime/runtime.js:271:22)
>     at Generator.prototype.(anonymous function) [as throw] (/home/X/back/node_modules/regenerator-runtime/runtime.js:97:21)
>     at tryCatch (/home/X/back/node_modules/regenerator-runtime/runtime.js:45:40)
>     at invoke (/home/X/back/node_modules/regenerator-runtime/runtime.js:135:20)
>     at /home/X/back/node_modules/regenerator-runtime/runtime.js:147:13
>     at <anonymous>
>     at process._tickCallback (internal/process/next_tick.js:189:7)
>      errors: { 'name.translations': [Object], name: [Object] },
>      _message: 'Language validation failed',
>      name: 'ValidationError' }

I've printed the data I'm trying to save : 

```
{
     "_id": "aaaaa0000000000000000000",
     "name": {
          "_id": "5e0f30f2c38f2b59775d9bb3",
          "translations": [
               {
                    "idLanguage": "aaaaa0000000000000000000",
                    "value": "Français"
               }
          ],
          "description": "French"
     },
     "acronym": "FR",
     "__v": 0
}
```

It should work.

**PS**: The exact same code worked in previous version of mongoose 5.6.x

**What is the expected behavior?**

The document get successfully saved

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

> node.js v8.16.2
> mongodb v4.2.1
> mongoose v5.8.4 (also tried with v5.7.0)

**Comment by GregoryNEUT** [comment]
Without changing the data I want to save, I changed the schema to : 

 ```
{
        // Name of the language
        name: new Schema({
          description: {
            type: String,
          },

          translations: [  Schema.Types.Mixed  ],
        }),

        // Acronym that define the language (Ex: FR, EN, ...)
        acronym: {
          type: String,
          required: true,
        },
},
```

And it worked.

Here is the final data on the database : 

![image](https://user-images.githubusercontent.com/22965416/71730600-e9981980-2e42-11ea-8137-647c7e258da3.png)

Please find a proper soluce, it's only a trick

**Comment by vkarpov15** [comment]
The below script succeeds without throwing any errors, please modify the below script to demonstrate your issue.

```javascript
'use strict';
  
const mongoose = require('mongoose');

mongoose.set('useFindAndModify', false);

const { Schema } = mongoose;

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const schema = Schema({
    name: Schema({
      description: String,
      translations: [Schema({
        idLanguage: {
          type: Schema.Types.ObjectId,
          ref: 'Language',
        },
        value: String
      })]
    })
  });

  const Model = mongoose.model('Test', schema);

  await Model.create({ name: { description: 'Test' } });

  const doc = await Model.findOne();
  doc.name.translations.push({
    "idLanguage": "aaaaa0000000000000000000",
    "value": "Français"
  });

  await doc.save();

  console.log('Done');
}

```

Also, it looks like you're using some sort of transpiler, can you please clarify what transpiler you're using?

**Comment by GregoryNEUT** [comment]
I've inserted your code inside of my project and it works. Still doesn't with my specific case inside of the same project. Here is the _console.log_ of the model I'm trying to save.

I'll investigate and try to see what I'm doing differently and will comes back to you. Thx for your time.

Also, I'm using Babel transpiler : 

>     "@babel/cli": "^7.4.4",
>     "@babel/core": "^7.4.4",
>     "@babel/plugin-proposal-class-properties": "^7.5.5",
>     "@babel/plugin-proposal-object-rest-spread": "^7.5.5",
>     "@babel/plugin-proposal-private-methods": "^7.4.4",
>     "@babel/plugin-transform-runtime": "^7.4.4",
>     "@babel/plugin-transform-spread": "^7.2.2",
>     "@babel/preset-env": "^7.4.4",
>     "@babel/preset-typescript": "^7.6.0",

**.babelrc**

```
{
  "presets": [
    "@babel/env",
    "@babel/typescript"
  ],
  "plugins": [
    "@babel/transform-runtime",
    "@babel/plugin-proposal-optional-chaining",
    "@babel/plugin-proposal-private-methods",
    [
      "@babel/plugin-proposal-decorators",
      {
        "decoratorsBeforeExport": true
      }
    ],
    "@babel/proposal-class-properties",
    [
      "@babel/plugin-transform-spread",
      {
        "loose": true
      }
    ],
    "@babel/plugin-proposal-nullish-coalescing-operator",
    [
      "module-resolver",
      {
        "cwd": "babelrc",
        "alias": {
          "@eliot_backend": "./src"
        }
      }
    ]
  ]
}

```

```
    $__: {
        strictMode: true,
        selected: {
        },
        shardval: null,
        saveError: null,
        validationError: {
            errors: {
                name.translations: {
                    stringValue: '"[ { idLanguage: 'aaaaa0000000000000000000', value: 'Français' } ]"',
                    kind: 'Array',
                    value: [
                        {
                            idLanguage: 'aaaaa0000000000000000000',
                            value: 'Français',
                        },
                    ],
                    path: 'translations',
                    reason: {
                    },
                    message: 'Cast to Array failed for value "[ { idLanguage: 'aaaaa0000000000000000000', value: 'Français' } ]" at path "translations"',
                    name: 'CastError',
                },
            },
            _message: 'Language validation failed',
            message: 'Language validation failed: name.translations: Cast to Array failed for value "[ { idLanguage: 'aaaaa0000000000000000000', value: 'Français' } ]" at path "translations"',
            name: 'ValidationError',
        },
        adhocPaths: null,
        removing: null,
        inserting: null,
        saving: null,
        version: null,
        getters: {
        },
        _id: 'aaaaa0000000000000000000',
        populate: null,
        populated: null,
        wasPopulated: false,
        scope: null,
        activePaths: {
            paths: {
                acronym: 'init',
                _id: 'init',
                name: 'init',
                __v: 'init',
            },
            states: {
                ignore: {
                },
                default: {
                },
                init: {
                    _id: true,
                    name: true,
                    acronym: true,
                    __v: true,
                },
                modify: {
                },
                require: {
                },
            },
            stateNames: [
                'require',
                'modify',
                'init',
                'default',
                'ignore',
            ],
        },
        pathsToScopes: {
        },
        cachedRequired: {
        },
        session: null,
        $setCalled: {
        },
        ownerDocument: null,
        fullPath: null,
        emitter: {
            domain: null,
            _events: {
                save: [
                    function () {
        _this.emit('save', _this);
        _this.constructor.emit('save', _this);
      },
                    function () {
        _this.emit('save', _this);
        _this.constructor.emit('save', _this);
      },
                ],
                isNew: [
                    function (val) {
        _this.isNew = val;
        _this.emit('isNew', val);
        _this.constructor.emit('isNew', val);
      },
                    function (val) {
        _this.isNew = val;
        _this.emit('isNew', val);
        _this.constructor.emit('isNew', val);
      },
                ],
            },
            _eventsCount: 2,
            _maxListeners: 0,
        },
        $options: {
            skipId: true,
            isNew: false,
            willInit: true,
        },
    },
    isNew: false,
    errors: [Circular -> errors],
    _doc: {
        _id: 'aaaaa0000000000000000000',
        name: {
            $parent: [Circular -> $$ROOT],
            $isSingleNested: true,
            _doc: {
                _id: '5e12e727d5c2d952cd9b43f4',
                translations: [
[Circular -> $$ROOT],
                ],
                description: 'French',
            },
            $__: {
                strictMode: true,
                selected: {
                },
                shardval: null,
                saveError: null,
                validationError: {
                    errors: {
                        translations: [Circular -> name.translations],
                    },
                    _message: 'Validation failed',
                    message: 'Validation failed: translations: Cast to Array failed for value "[ { idLanguage: 'aaaaa0000000000000000000', value: 'Français' } ]" at path "translations"',
                    name: 'ValidationError',
                },
                adhocPaths: null,
                removing: null,
                inserting: null,
                saving: null,
                version: null,
                getters: {
                },
                _id: '5e12e727d5c2d952cd9b43f4',
                populate: null,
                populated: null,
                wasPopulated: false,
                scope: null,
                activePaths: {
                    paths: {
                    },
                    states: {
                        ignore: {
                        },
                        default: {
                        },
                        init: {
                        },
                        modify: {
                        },
                        require: {
                        },
                    },
                    stateNames: [
                        'require',
                        'modify',
                        'init',
                        'default',
                        'ignore',
                    ],
                },
                pathsToScopes: {
                },
                cachedRequired: {
                },
                session: null,
                $setCalled: {
                },
                ownerDocument: [Circular -> $$ROOT],
                fullPath: null,
                emitter: {
                    domain: null,
                    _events: {
                    },
                    _eventsCount: 0,
                    _maxListeners: 0,
                },
                $options: {
                    priorDoc: {
                        $parent: [Circular -> $$ROOT],
                        $isSingleNested: true,
                        $__: {
                            strictMode: true,
                            selected: {
                            },
                            shardval: null,
                            saveError: null,
                            validationError: null,
                            adhocPaths: null,
                            removing: null,
                            inserting: null,
                            saving: null,
                            version: null,
                            getters: {
                            },
                            _id: '5e12e727d5c2d952cd9b43f4',
                            populate: null,
                            populated: null,
                            wasPopulated: false,
                            scope: null,
                            activePaths: {
                                paths: {
                                    _id: 'init',
                                    translations: 'init',
                                    description: 'init',
                                },
                                states: {
                                    ignore: {
                                    },
                                    default: {
                                    },
                                    init: {
                                        _id: true,
                                        description: true,
                                        translations: true,
                                    },
                                    modify: {
                                    },
                                    require: {
                                    },
                                },
                                stateNames: [
                                    'require',
                                    'modify',
                                    'init',
                                    'default',
                                    'ignore',
                                ],
                            },
                            pathsToScopes: {
                            },
                            cachedRequired: {
                            },
                            session: null,
                            $setCalled: {
                            },
                            ownerDocument: [Circular -> $$ROOT],
                            fullPath: null,
                            emitter: {
                                domain: null,
                                _events: {
                                },
                                _eventsCount: 0,
                                _maxListeners: 0,
                            },
                            $options: {
                                isNew: false,
                            },
                        },
                        isNew: false,
                        errors: null,
                        _doc: {
                            _id: '5e12e727d5c2d952cd9b43f4',
                            translations: [
[Circular -> $$ROOT],
                            ],
                            description: 'French',
                        },
                        $locals: {
                        },
                        $init: true,
                    },
                    isNew: false,
                },
            },
            isNew: false,
            errors: [Circular -> errors],
            $locals: {
            },
        },
        acronym: 'FR',
        __v: 0,
    },
    $locals: {
    },
    $init: true,
}

```

**Comment by GregoryNEUT** [comment]
@vkarpov15 

First, you are right I'm using a transpiler: Babel for my main project and typescript for a small package I use to interact with the database _(schema handler)_.

I've made a lot of different try and I have state that : 

- The code you have provided me works in my project
- I can make my code work using Schema.Type.Mixed
- I can make my code work if I build the entire Schema in the main project (in opposition with building the Schema in my lower level package)

**My conclusion** : 

It's 100% a transpiler issue, I got no issue when I perform every mongoose action at the same location.


Thank you for your help, I think this issue is VERY specific. Indeed my main project is migrating to typescript so I am using babel, when my lower packages are already in typescript _(compilation target setted to es5)_.

I'll set my Schema to **Schema.Type.Mixed** until I finish the migration of my whole project into typescript.
